compiler/lib/specialize_js.ml

```diff
@@ -22,49 +22,51 @@ open! Stdlib
 open Code
 open Flow
 
-let specialize_instr info i =
-  match i with
-  | Let (x, Prim (Extern "caml_format_int", [ y; z ])) -> (
-      match the_string_of info y with
+let specialize_instr ~target info i =
+  match i, target with
+  | Let (x, Prim (Extern "caml_format_int", [ y; z ])), `JavaScript -> (
+      match the_string_of ~target info y with
       | Some "%d" -> (
-          match the_int info z with
+          match the_int ~target info z with
           | Some i -> Let (x, Constant (String (Int32.to_string i)))
           | None -> Let (x, Prim (Extern "%caml_format_int_special", [ z ])))
       | _ -> i)
-  | Let (x, Prim (Extern "%caml_format_int_special", [ z ])) -> (
-      match the_int info z with
+  | Let (x, Prim (Extern "%caml_format_int_special", [ z ])), `JavaScript -> (
+      match the_int ~target info z with
       | Some i -> Let (x, Constant (String (Int32.to_string i)))
       | None -> i)
   (* inline the String constant argument so that generate.ml can attempt to parse it *)
-  | Let
-      ( x
-      , Prim
-          ( Extern (("caml_js_var" | "caml_js_expr" | "caml_pure_js_expr") as prim)
-          , [ (Pv _ as y) ] ) )
+  | ( Let
+        ( x
+        , Prim
+            ( Extern (("caml_js_var" | "caml_js_expr" | "caml_pure_js_expr") as prim)
+            , [ (Pv _ as y) ] ) )
+    , _ )
     when Config.Flag.safe_string () -> (
```

Comment:
Do you want this with wasm ?

Reply:
Do you mean that this flag should only matter when emitting JS?

Reply:
Fixed

Reply:
I think we don't care at all about safe_string here. `Flow.the_string_of` is supposed to do the work already.

Reply:
(Tangent: Is `safe_string` about string mutability? And if yes, since it has been dropped since 4.02, maybe we could drop this flag as well?)

Reply:
I think it's 4.10

> OCaml 4.10.0 (21 February 2020)
> -------------------------------
> ### Language features
> ...
> * #1859, #9117: enforce safe (immutable) strings by removing
  the -unsafe-string option by default. This can be overridden by
  a configure-time option (available since 4.04 in 2016):
  --disable-force-safe-string since 4.08, -no-force-safe-since
  between 4.07 and 4.04.
  In the force-safe-string mode (now the default), the return type of the
  String_val macro in C stubs is `const char*` instead of
  `char*`. This change may break C FFI code.
  (Kate Deplaix)

Reply:
or even OCaml 5 for 
> * #10893: Remove configuration options --disable-force-safe-string and
  DEFAULT_STRING=unsafe as well as --enable-force-safe-string and
  DEFAULT_STRING=safe which are now the default unconditionally.
  (Kate Deplaix, review by Gabriel Scherer and David Allsopp)